# JS Promise

Promises can either fail or Succeed. Depending on the Promise failing or succeeding the corresponding callback method is called.

This happens independent of when the promise is resolved. This is very helpfull when programming asynchronous, as you can react to the promises outcome as soon as it's ready.

This is how you define a Promise:

````javascript
var promise = new Promise(function(resolve, reject) {
  // do a thing, possibly async, thenâ€¦

  if (/* everything turned out fine */) {
    resolve("Stuff worked!");
  }
  else {
    reject(Error("It broke"));
  }
});
````

The promise constructor takes one argument, a callback with two parameters, resolve and reject. Do something within the callback, perhaps async, then call resolve if everything worked, otherwise call reject.

This is how you use a promise:

````javascript
promise.then(function(result) {
  console.log(result); // "Stuff worked!"
}, function(err) {
  console.log(err); // Error: "It broke"
});
````

